plugin/storage/es/samplingstore/storage.go

```diff
@@ -0,0 +1,189 @@
+	throughputType       = "throughput-sampling"
+	probabilitiesType    = "probabilities-sampling"
+	indexPrefixSeparator = "-"
+)
+
+type SamplingStore struct {
+	client              func() es.Client
+	logger              *zap.Logger
+	samplingIndexPrefix string
+	indexDateLayout     string
+	maxDocCount         int
+}
+
+type SamplingStoreParams struct {
+	Client          func() es.Client
+	Logger          *zap.Logger
+	IndexPrefix     string
+	IndexDateLayout string
+	MaxDocCount     int
+}
+
+func NewSamplingStore(p SamplingStoreParams) *SamplingStore {
+	return &SamplingStore{
+		client:              p.Client,
+		logger:              p.Logger,
+		samplingIndexPrefix: prefixIndexName(p.IndexPrefix, samplingIndex),
+		indexDateLayout:     p.IndexDateLayout,
+		maxDocCount:         p.MaxDocCount,
+	}
+}
+
+func (s *SamplingStore) InsertThroughput(throughput []*model.Throughput) error {
+	ts := time.Now()
+	writeIndexName := indexWithDate(s.samplingIndexPrefix, s.indexDateLayout, ts)
+	s.writeThroughput(writeIndexName, ts, throughput)
+	return nil
+}
+
+func (s *SamplingStore) InsertProbabilitiesAndQPS(hostname string,
+	probabilities model.ServiceOperationProbabilities,
+	qps model.ServiceOperationQPS,
+) error {
+	ts := time.Now()
+	writeIndexName := indexWithDate(s.samplingIndexPrefix, s.indexDateLayout, ts)
+	val := dbmodel.ProbabilitiesAndQPS{
+		Hostname:      hostname,
+		Probabilities: probabilities,
+		QPS:           qps,
+	}
+	s.writeProbabilitiesAndQPS(writeIndexName, ts, val)
+	return nil
+}
+
+func (s *SamplingStore) writeProbabilitiesAndQPS(indexName string, ts time.Time, pandqps dbmodel.ProbabilitiesAndQPS) {
+	s.client().Index().Index(indexName).Type(probabilitiesType).
+		BodyJson(&dbmodel.TimeProbabilitiesAndQPS{
+			Timestamp:           ts,
+			ProbabilitiesAndQPS: pandqps,
+		}).Add()
+}
+
+func indexWithDate(indexNamePrefix, indexDateLayout string, date time.Time) string {
+	return indexNamePrefix + date.UTC().Format(indexDateLayout)
+}
+
+func (s *SamplingStore) writeThroughput(indexName string, ts time.Time, throughputs []*model.Throughput) {
+	s.client().Index().Index(indexName).Type(throughputType).
+		BodyJson(&dbmodel.TimeThroughput{
+			Timestamp:  ts,
+			Throughput: dbmodel.FromThroughputs(throughputs),
+		}).Add()
+}
+
+func (s *SamplingStore) GetThroughput(start, end time.Time) ([]*model.Throughput, error) {
+	ctx := context.Background()
+	indices := s.getReadIndices(start, end)
+	searchResult, err := s.client().Search(indices...).
+		Size(s.maxDocCount).
+		Query(buildTSQuery(start, end)).
+		IgnoreUnavailable(true).
+		Do(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search for throughputs: %w", err)
+	}
+	var retSamples []dbmodel.Throughput
+	hits := searchResult.Hits.Hits
+	for _, hit := range hits {
+		source := hit.Source
+		var tToD dbmodel.TimeThroughput
+		if err := json.Unmarshal(*source, &tToD); err != nil {
+			return nil, errors.New("unmarshalling ElasticSearch documents failed")
```

Comment:
don't swallow errors
```suggestion
			return nil, errors.New("unmarshalling ElasticSearch documents failed: %w", err)
```

Reply:
don't need ElasticSearch in the message